### PR TITLE
[RU] Add Russian translation to /specifics/ecto.md

### PR DESCRIPTION
**Pull request by tetiana12345678**

### CODE REVIEW

**Comment by pjhampton** [comment]
Thanks @tetiana12345678. Can you review this please @brain-geek? :+1:

**Comment by doomspork** [comment]
@brain-geek when this is good to go, feel free to merge this yourself :+1:

File: ru/lessons/specifics/ecto.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+category: specifics
+order: 2
+lang: ru
+---
+
+Ecto является официальным Elixir проектом. Это оболочка, которая предоставляет возможность коммунникации с базой данных. Ecto позволяет создавать миграции, обьявлять модели, вносить и обновлять данные, а также посылать запросы к ним.
+
+## Содержание
+
+- [Установка](#setup)
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: You've forgotten to fix nav links

File: ru/lessons/specifics/ecto.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
+---
+layout: page
+title: Ecto
+category: specifics
+order: 2
+lang: ru
+---
+
+Ecto является официальным Elixir проектом. Это оболочка, которая предоставляет возможность коммунникации с базой данных. Ecto позволяет создавать миграции, обьявлять модели, вносить и обновлять данные, а также посылать запросы к ним.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: `Ecto является официальным Elixir проектом` - > `Ecto - официальный проект от создателей Elixir/официальный подпроект`?

File: ru/lessons/specifics/ecto.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+Создадим репозиторий для проекта - оболочку для базы данных. Для этого выполним команду`mix ecto.gen.repo`. Репозиторий находится в `lib/<project name>/repo.ex`.
+
+```elixir
+defmodule ExampleApp.Repo do
+  use Ecto.Repo,
+    otp_app: :example_app
+end
+```
+
+### Супервайзер
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Please also add this term to glossary (in the root of project)

File: ru/lessons/specifics/ecto.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+После создания репозитория, необходимо создать иерархию наблюдателей. Файл находится в `lib/<project name>.ex`.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: I've found alternative existing translation for the "supervisor tree" term in the Armstrong translation book - `дерево надзора`. https://github.com/dyp2000/Russian-Armstrong-Erlang/blob/master/ru/18/Programming_Erlang_Ch18.md + same term used here http://erlanger.ru/wiki/index.php/OTP_Applications

Looks like that's already standard translation

File: ru/lessons/specifics/ecto.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+Важно, чтоб мы настроили Репозиторий, как супервайзер с помощью `supervisor/3` метода, а не `worker/3`. Если создать приложение с флагом `--sup`, большинство нужного кода для нас уже будет сгенерировано.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Do we need to have `Репозиторий` starting with upcase letter?

File: ru/lessons/specifics/ecto.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    children = [
+      supervisor(ExampleApp.Repo, [])
+    ]
+
+    opts = [strategy: :one_for_one, name: ExampleApp.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+Чтоб узнать больше о супервайзерах советуем посетить урок [OTP Supervisors](/lessons/advanced/otp-supervisors).
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Name of the lesson can be translated too.

File: ru/lessons/specifics/ecto.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    end
+
+    create unique_index(:users, [:username], name: :unique_usernames)
+  end
+end
+```
+
+Ecto по умолчанию создает первичный автоинкрементируемый ключ под названием id. Так же по умолчанию используется `change/0` callback, но для большего контроля можно использовать `up/0` и `down/0`.
+
+Вы уже могли догадаться, что при добавлении поля `timestamps` к миграции у Вас появляется возможность создать и управлять полями `created_at` и `updated_at`.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: `добавлении поля `timestamps` к миграции` `As you might have guessed, adding timestamps to your migration`

'timestamp' is not field in this context, but part of the DSL

File: ru/lessons/specifics/ecto.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+Следующим шагом будет запуск только что созданной миграции с помощью команды `mix ecto.migrate`.
+
+Дополнительную информацию о миграциях Вы можете найти в главе [Ecto.Migration](http://hexdocs.pm/ecto/Ecto.Migration.html#content).
+
+## Модели
+
+Теперь, имея миграцию, можно двигаться дальше, к модели. Модели определяют структуру(shema), вспомогательныe методы(helper methods) и изменяемые наборы. Изменяемые наборы обьяснены в следующей главе.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: shema -> schema 
`изменяемые наборы` does not sound very good, while yet I don't see better term.

File: ru/lessons/specifics/ecto.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+  @optional_fields ~w()
+
+  def changeset(user, params \\ :empty) do
+    user
+    |> cast(params, @required_fields, @optional_fields)
+    |> unique_constraint(:username)
+  end
+end
+```
+Структура(schema), заданная в нашей модели, очень близка к коду в самой миграции. Дополнительно к полям базы данных мы также включили 2 виртуальных поля. Виртуальные поля не сохраняются в базу данных, но могут быть полезными для валидаций. Примеры виртуальных полей можно найти здесь [Changesets](#changesets).
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: `очень близка к коду` -> `похожа`?
включили -> добавили ?

**Comment by brain-geek** [comment]
Btw, about the 'changeset' translation. As for me, 'Изменяемые наборы' sounds like 'changable sets'(aka set which can be changed), while 'Набор изменений' sounds a lot more like 'changeset'(set of changes).

File: ru/lessons/specifics/ecto.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    add_error(changeset, :password, "is not valid")
+  end
+end
+```
+
+Таким образом мы улучшили функцию `changeset/2`, а также добавили три новые вспомогательные функции: `validate_password_confirmation/1`, `password_mismatch_error/1`, и `password_incorrect_error/1`.
+
+По названию можно догадаться, что функция `changeset/2` создает новый изменяемый набор. В теле функции используем функцию `cast/4`, которая трансформирует данные параметры в изменяемый набор, который состоит из обязатаельных и необязательных полей. Далее происходит валидация длинны пароля, находящегося в изменяемом наборе. После этого, используя созданную нами фукнцию, мы можем проверить, что подтверждение пароля является таким же как и сам пароль, а также проверить уникальность имени пользователя. В конце, обновим пароль в базе данных. Для этого используем функцию `put_change/3`, которая обновит значение в именяемом наборе.
+
+Использование `User.changeset/2` довольно просто:
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Использование -> использовать?

File: ru/lessons/specifics/ecto.md (hunk 4 above)
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: In this lesson "supervisor" is translated as "супервайзер", while in #239 the translation in both the lesson and the glossary is "супервизор". Also "супервизор" is used in https://github.com/dyp2000/Russian-Armstrong-Erlang/blob/master/ru/18/Programming_Erlang_Ch18.md . I guess we should decide which one to use and stick to it in these and further translations.

Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: good point. I have seen/heard people using word `супервайзер`, but haven't heard people saying/writing `супервизор`, that's why I went with the first version. I am happy to change it if we decide to do so.

File: ru/lessons/specifics/ecto.md (hunk 1 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: :+1: thanks. Will address all of the comments in free time after work

File: ru/lessons/specifics/ecto.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+  defp password_incorrect_error(changeset) do
+    add_error(changeset, :password, "is not valid")
+  end
+end
+```
+
+Таким образом мы улучшили функцию `changeset/2`, а также добавили три новые вспомогательные функции: `validate_password_confirmation/1`, `password_mismatch_error/1`, и `password_incorrect_error/1`.
+
+По названию можно догадаться, что функция `changeset/2` создает новый изменяемый набор. В теле функции используем функцию `cast/4`, которая трансформирует данные параметры в изменяемый набор, который состоит из обязатаельных и необязательных полей. Далее происходит валидация длинны пароля, находящегося в изменяемом наборе. После этого, используя созданную нами фукнцию, мы можем проверить, что подтверждение пароля является таким же как и сам пароль, а также проверить уникальность имени пользователя. В конце, обновим пароль в базе данных. Для этого используем функцию `put_change/3`, которая обновит значение в именяемом наборе.
```
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: There are several typos on this line:
`обязатаельных` → `обязательных`,
`длинны` → длины`
`фукнцию` → `функцию`
`именяемом наборе` → `наборе изменений`.

File: ru/lessons/specifics/ecto.md (hunk 4 above)
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Looks like both terms are used in the wild in different communities. I think that we should use the one already adopted widely

One more thing to think about: google://"site:erlanger.ru супервайзер" says no results, but google://"site:erlanger.ru супервизор" has 6. 

Do we have any more russian-speaking erlang communities/documents/books to allign to?

Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Looks like both terms are used in the wild in different communities. I think that we should use the one already adopted widely

One more thing to think about: google://"site:erlanger.ru супервайзер" says no results, but google://"site:erlanger.ru супервизор" has 6. 

Do we have any more russian-speaking erlang communities/documents/books to allign to?

File: ru/lessons/specifics/ecto.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+    add_error(changeset, :password_confirmation, "Passwords does not match")
+  end
+
+  defp password_incorrect_error(changeset) do
+    add_error(changeset, :password, "is not valid")
+  end
+end
+```
+
+Таким образом мы улучшили функцию `changeset/2`, а также добавили три новые вспомогательные функции: `validate_password_confirmation/1`, `password_mismatch_error/1`, и `password_incorrect_error/1`.
```
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: If i am not mistaken, there should be a comma after "Таким образом" but not after `password_mismatch_error/1`.

File: ru/lessons/specifics/ecto.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+
+Примеры построения различных запросов можно найти здесь [Ecto.Query.API](http://hexdocs.pm/ecto/Ecto.Query.API.html).
+
+## Изменяемые наборы
+
+В предыдущем параграфе мы увидели как извлечь необходимую информацию из базы данных, а что на счет добавления и обновления записей? Для этого нам необходимы изменяемые наборы (Changesets).
+
+Изменяемые наборы выполняют функции фильтрации, валидации, и поддержки ограничений при изменении модели.
+
+Давайте используем пример изменяемых наборов для создания аккаута для пользователя. Для начала необходимо обновить модель:
```
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: `аккаута` → `аккаунта`?

File: ru/lessons/specifics/ecto.md (hunk 4 above)
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: I've also found another translation option: `?site:erlanger.ru супервайзор` &mdash; 4 results.
Results from another resource:
`?site:erlang-russian.org супервизор` &mdash; 3 results,
`?site:erlang-russian.org супервайзер` &mdash; 2 results,
`?site:erlang-russian.org супервайзор` &mdash; none.

File: ru/lessons/specifics/ecto.md (hunk 8 above)
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: ` обьяснены` → `объяснены`.

File: ru/lessons/specifics/ecto.md (hunk 2 above)
Comment (nscyclone) [inline ru/lessons/specifics/ecto.md]: ` коммунникации` → `коммуникации`.

File: ru/lessons/specifics/ecto.md (hunk 4 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: so, which one shall we stick to? `супервизор` seem to be most widely used from your research.

File: ru/lessons/specifics/ecto.md (hunk 8 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: `изменяемые наборы` term open for discussion. That was my creative thinking. That's the best I can come up with. Happy to change if we find better translation for `changesets`

File: ru/lessons/specifics/ecto.md (hunk 11 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: o dear. thank you for noticing. I was getting a bit tired closer to the end of the article. :sleepy:

File: ru/lessons/specifics/ecto.md (hunk 10 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: I think you can use either here.

**Comment by tetiana12345678** [comment]
Before I go ahead and make changes throughout the file, can we agree on 2 terms:
`changeset` -> `набор изменений` ?
`supervision tree` -> `дерево надзора`
`supervisor` -> `супервизор` or `супервайзер` or `супервайзор`.

**Comment by brain-geek** [comment]
`changeset -> набор изменений` <- I think this one is good
`supervisor` - супервизор, I guess. @nscyclone , what's your opinion on this?

**Comment by nscyclone** [comment]
I'd agree with @brain-geek , in my opinion, `набор изменений` and `супервизор` are the most fitting here.

File: ru/lessons/specifics/ecto.md (hunk 12 above)
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: I agree about both. Proof about the first one: http://new.gramota.ru/spravka/punctum?layout=item&id=58_715

**Comment by tetiana12345678** [comment]
@nscyclone @brain-geek updated :smile_cat:

File: GLOSSARY.md [hunk 14]
```diff
@@ -17,6 +17,7 @@ Class names and proper nouns do not get translated.
 - function clauses - объявления функций
 - guard - ограничитель
 - guard operator - ограничивающий оператор
+- changesets - набор изменений
```
Comment (brain-geek) [inline GLOSSARY.md]: `changeset` should be plural, I guess.

File: ru/lessons/specifics/ecto.md [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+
+      timestamps
+    end
+
+    create unique_index(:users, [:username], name: :unique_usernames)
+  end
+end
+```
+
+Ecto по умолчанию создает первичный автоинкрементируемый ключ под названием id. Так же по умолчанию используется `change/0` callback, но для большего контроля можно использовать `up/0` и `down/0`.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: `Так же` should be 'Также` - http://tak-zhe.ru
> Также — аналог союза «и», словосочетания «к тому же».

File: ru/lessons/specifics/ecto.md [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+
+Вы уже могли догадаться, что при добавлении `timestamps` к миграции у Вас появляется возможность создать и управлять полями `created_at` и `updated_at`.
+
+Следующим шагом будет запуск только что созданной миграции с помощью команды `mix ecto.migrate`.
+
+Дополнительную информацию о миграциях Вы можете найти в главе [Ecto Миграции](http://hexdocs.pm/ecto/Ecto.Migration.html#content).
+
+## Модели
+
+Теперь, имея миграцию, можно двигаться дальше, к модели. Модели определяют структуру(schema), вспомогательныe методы(helper methods) и набор изменений. Набор изменений объяснен в следующей главе.
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Looks like we should add `schema` to glossary

File: ru/lessons/specifics/ecto.md [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+  @optional_fields ~w()
+
+  def changeset(user, params \\ :empty) do
+    user
+    |> cast(params, @required_fields, @optional_fields)
+    |> unique_constraint(:username)
+  end
+end
+```
+Структура(schema), заданная в нашей модели, очень похожа на код в самой миграции. Дополнительно к полям базы данных мы также добавили 2 виртуальных поля. Виртуальные поля не сохраняются в базу данных, но могут быть полезными для валидаций. Примеры виртуальных полей можно найти здесь [Набор изменений](#section-15).
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Do we really need explanation `(schema)` multiple times in one document.

File: ru/lessons/specifics/ecto.md [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+query = from u in User,
+    where: fragment("downcase(?)", u.username) == ^username
+    select: u
+```
+
+Примеры построения различных запросов можно найти здесь [Ecto.Query.API](http://hexdocs.pm/ecto/Ecto.Query.API.html).
+
+## Набор изменений
+
+В предыдущем параграфе мы увидели как извлечь необходимую информацию из базы данных, а что на счет добавления и обновления записей? Для этого нам необходим набор изменений (Changesets).
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: `на счет` looks more like `насчет`. http://rus.stackexchange.com/a/154
> Все зависит от контекста. Но проще всего проверять так: если между "на" и "счет" можно вставить слово, то пишется раздельно. И наоборот.

**Comment by brain-geek** [comment]
@tetiana12345678 Thanks!

I've looked through once again and found some more typos.

File: GLOSSARY.md (hunk 14 above)
Comment (tetiana12345678) [inline GLOSSARY.md]: @brain-geek thanks for looking at PR again. I don't quite understand what you mean by `changeset should be plural, I guess`. Are you talking about russian translation to be `наборы изменений`?

Comment (brain-geek) [inline GLOSSARY.md]: I am talking about english wording here. E.g. tuple, guard, sigil but changeset**s**.

Comment (tetiana12345678) [inline GLOSSARY.md]: ow, I see. Change is coming :smiley:

**Comment by tetiana12345678** [comment]
updated :smiley:

**Comment by doomspork** [comment]
@tetiana12345678 / @brain-geek is this ready for merge?

**Comment by tetiana12345678** [comment]
If @brain-geek happy to merge it :smile:

File: ru/lessons/specifics/ecto.md [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+
+Ecto - официальный проект от создателей Elixir. Это оболочка, которая предоставляет возможность коммуникации с базой данных. Ecto позволяет создавать миграции, объявлять модели, вносить и обновлять данные, а также посылать запросы к ним.
+
+## Содержание
+
+- [Установка](#section-1)
+  - [Репозиторий](#section-2)
+  - [Супервизор](#section-3)
+  - [Настройка](#section-4)
+- [Mix задачи](#section-5)
```
Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Most likely this one link is broken

**Comment by brain-geek** [comment]
@tetiana12345678 Can you fix the nav? Everything else looks good

File: ru/lessons/specifics/ecto.md (hunk 19 above)
Comment (tetiana12345678) [inline ru/lessons/specifics/ecto.md]: Not sure I know how. Any suggestion on how to fix it? @brain-geek

Comment (brain-geek) [inline ru/lessons/specifics/ecto.md]: Just run application locally (`bundle exec jekyll s`, I guess) and check the link correctness.

Ping me if you need any help on that!

**Comment by tetiana12345678** [comment]
@brain-geek links are all working. I have tested them 😸

**Comment by brain-geek** [comment]
Thanks, @tetiana12345678 !

**Pull request merged**